Method for cameraoptions from camera and center

BAschl opened the pull request:
As requested in #1414 I created a method to calculate the cameraoptions from a given center and camera, which can then be used in jumpTo, flyTo and so on.


Sadly it's not easy to know if the new center would be visible (or if there's a mountain inbetween), as this requires a redraw of the coordsframebuffer. Also note, that the returned cameraOptions could be beyond the constraints.

## Launch Checklist
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs. (You mean doc comments? If so, then yes)
 - [x] Manually test the debug page.
 - [x] Suggest a changelog category: feature
 - [x] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

HarelM commented:
Thanks for taking the time to do this!
Why is API receiving Mercator coordinates and not Lat lng?
Also there's a need to add change log entry in changelig.md.

BAschl commented:
Thank you.

It doesn't really matter. If it would take LngLat there would be the need for a 3rd parameter: cameraAltitude.
Should I change it to LngLat?

HarelM commented:
The general API is around lat lng, so I prefer to keep it this way. I'm not even sure if mercator coordinates is actually exported publicly at this point, so I would feel better if the API will remain with latlng, and latitude.
I know this makes the call signature a bit more dirty, but it avoids the need to use the mercator class just before calling this method. IDK...

BAschl commented:
Okay, done :)

HarelM commented on src/ui/camera.ts:
```diff
@@ -775,6 +776,37 @@ abstract class Camera extends Evented {
+     * @param to the center to look at
+     * @returns {CameraOptions} the calculated camera options
+     */
+    calculateCameraOptionsFromTo(from: LngLat, altitudeFrom:number, to: LngLat) : CameraOptions {
```
Do we assume we are always looking at the ground? Can we have an optional parameter as a second altitude that defaults to 0 (or maybe the ground height for 3D)?
Also missing a space - i.e should be `altitudeFrom: number`

BAschl replied:
Oops, you are right.
I'll override the method in map to get the elevation if no target altitude was given and then call the super method.

Did you mean we should also calculate a new center if toAltitude is given? Would be fairly simple without terrain. How would I go about this if there's terrain?

HarelM commented on src/ui/camera.test.ts:
```diff
@@ -55,6 +56,39 @@ function assertTransitionTime(done, camera, min, max) {
+    test('pitch 45', () => {
+        const cam = MercatorCoordinate.fromLngLat({lng: 1, lat: 0});
+        //up as far as away => pitch should be 45°
+        cam.z = cam.x - 0.5;
```
Not sure I understand this - can you clarify how we use x minus half a meter and create altitude that is 45 degrees?
Maybe a drawing would help me here.
I understand that it's probably correct, I'm just not sure why...

prozessor13 commented:
Thanks for your PR! Very useful :D.

I created a route-flight test-page which uses your method. Basically it works great when using it with `jumpTo`, but when using `flyTo` or `easeTo` the positioning is not correct, which surprised me a lot. Today i do not have time to digg more into this, so for now i send only my test-page. More tests will follow.

Regrads. Max
[route.zip](https://github.com/maplibre/maplibre-gl-js/files/9259018/route.zip)
.

prozessor13 commented:
When comment out:

- https://github.com/maplibre/maplibre-gl-js/blob/main/src/ui/camera.ts#L901
- https://github.com/maplibre/maplibre-gl-js/blob/main/src/ui/camera.ts#L937

the `flyTo` and `easeTo` works as expected and the route-flight looks much more smooth and natural. Sadly i cant remember why i added the `freezeElevation` logic into camera.ts, so i have to rethink the *freeze* logic!

prozessor13 commented:
Another interesting usecase of this method would be:

When the camera goes below terrain (f.e. during dragging), rise the camera above terrain, but let it look onto the same center-point.

HarelM commented:
I think the freeze elevation was due to the fact that the camera was "giggling" when you moved the map around (i.e. the camera would go up when you go up a mountains and go down when you go down to a valley), I'm not sure though...

prozessor13 commented:
yes you are right, but the freeze logic is implemented already in handler_manager.ts which handles the user-dragging interactions.
I vaguely remember that the problem had something to do with the *swipe* feature, especially on mobile devices. Definitely some tests are needed for this!

prozessor13 commented:
created a new pull-request for the freezeElevation logic problem, so from my perspective this pull request is ready for merge.

https://github.com/maplibre/maplibre-gl-js/pull/1514

HarelM commented on src/ui/map.ts:
```diff
@@ -600,6 +600,13 @@ class Map extends Camera {
         return this._controls.indexOf(control) > -1;
     }
 
+    calculateCameraOptionsFromTo(from: LngLat, altitudeFrom: number, to: LngLat, altitudeTo?: number) : CameraOptions {
```
Can this be tested as well?

HarelM commented:
Also there's a med to resolve conflicts. Thanks for taking the time to write this!

BAschl commented on test/debug-pages/route.html:
```diff
@@ -0,0 +1,179 @@
+              }
+          });
+          map.flyTo(map.calculateCameraOptionsFromTo(camera.coord.toLngLat(), camera.altitude, coordinates[0]));
+          //   setTimeout(() => animate(), 2000);
```
I added to the debug page (click or ctrl+click for flyTo and jumpTo respectively) and I was annoyed by waiting for the animation to finish. I'll enabled it again with the next commit.

BAschl commented:
I'll take a look at it tomorrow. I've been to a festival the last week and need today to recharge.

BAschl commented:
Should be all done now.

I added test cases for calling it on map with a stubbed terrain and additionally I added test cases for the returned zoom. Could you please take a look at it? Should I put constants for expectedZoom?

Furthermore i covered an edge case (From the same as To => division by 0) by throwing an error and wrote a test for it.

HarelM commented on src/ui/map.ts:
```diff
@@ -603,6 +603,13 @@ class Map extends Camera {
     }
 
+    calculateCameraOptionsFromTo(from: LngLat, altitudeFrom: number, to: LngLat, altitudeTo?: number) : CameraOptions {
+        if (!altitudeTo && this.style.terrain) {
```
`altitudeTo` should be checked for undefined (or null) as `0` is a valid value here...

BAschl replied:
whoops, good catch!

BAschl commented:
What? My local Lint doesn't catch everything? :o

drwestco commented on src/ui/camera.ts:
```diff
@@ -774,6 +775,41 @@ abstract class Camera extends Evented {
     }
 
+    /**
+     * Calculates pitch, zoom and bearing for looking at @param newCenter with the camera position being @param newCenter
```
Documentation needs improvement. The summary refers to params that don't exist. It's confusing that 'from' is referred to as a camera, but 'to' is referred to as a center. If I'm understanding correctly, 'from' is where the camera is located; 'to' is where it's pointing.

HarelM replied:
Feel free to open a PR, it's easier to understand it this way :-) Docs can always be improved.
